### PR TITLE
feat(server): Import or Export Project Functionality

### PR DESCRIPTION
**Pull request by hexaforce**
# Overview
Import or Export Project Functionality

### About Export. 
The JSON format for export is as follows. It is broadly divided into three sections: `plugins`, `project`, and `scene`. For the scene, the **sceneJSON** is output using the Builder.
```
{
  "plugins": [
    { ... }
  ],
  "project": {
    ...
  },
  "scene": {
    ...
    "widgets": [
      { ... }
    ],
    "widgetAlignSystem": {
      ...
    },
    "story": {
      ...
      "pages": [
        {
          ...
          "blocks": [
            { ... }
          ],
        }
      ]
    },
    "nlsLayers": [
      { ... }
    ],
    "layerStyles": [
      { ... }
    ]
  }
}
```

For importing properties, a Type is required, but since the scene JSON does not include Type in its properties, we added an export flag to the Builder instance.
```diff
type Builder struct {
   ...
++   exportType bool
}
```
The output will look like this:
```
"property": {
  "default": {
    "text": "hogehoge"
  }
}
```
After adding Type, the output will be:
```
"property": {
  "default": {
    "text": {
      "type": "string",
      "value": "hogehoge"
    }
  }
}
```

### About Import
The IDs of each object will be imported with the same ID, so if they already exist in the database, they will be overwritten. However, new IDs will be created for the associated properties.

Currently, the plugins, layers, tags, and clusters in the scene are not supported.
```
"scene": {
  ...
  "plugins": {}, // not supported
  "layers": [], // not supported
  "tags": [], // not supported
  "clusters": [], // not supported
}
```
### About GraphQL
We have added ExportProject and ImportProject mutations to the Project.

ExportProject returns the JSON data as raw. The process of downloading it as a .reearth file is expected to be handled on the browser side.

ImportProject is designed to upload a .reearth file.

## What I've done
The uploaded files are now imported into the respective DB repositories:
* ImportProject() // => project
* ImportPlugins() // => plugin
* ImportScene() // => storytelling, property
* ImportNLSLayers() // => nlsLayer
* ImportStyles() // => style
* ImportStoryTelling() // => storytelling, property

I have implemented the corresponding tests:
* TestImportProject()
* TestImportPlugins()
* TestImportScene()
* TestImportNLSLayers()
* TestImportStyles()
* TestImportStoryTelling()

I have also implemented end-to-end tests for GraphQL:
* TestCallExportProject()
* TestCallImportProject()

I created a common function getWidgePlugin() in server/internal/usecase/interactor/scene.go to retrieve Widget plugins and refactored the code accordingly.

I created a common function ParseSceneJSON() for parsing the package. Since sceneJSON is private, I added this function to the builder package.

Additionally, I created the following functions:

* ParserWidgetAlignSystem() for parsing the complex WidgetAlignSystem
* AddItemFromPropertyJSON() for parsing and adding properties

All of these functions are located in server/pkg/scene/builder/decoder.go.

### Regarding Plugins

Plugins only retrieve data from the database. Plugin iles are not supported.

## How I tested

* Testing with test code
* Verifying functionality through web calls

## Memo
reference
https://www.notion.so/eukarya/Import-or-Export-Project-Functionality-728f84108ad54d8bb0d15b21fc3fa68e

## Release Notes

- **New Features**
  - Added GraphQL mutations for importing and exporting project data, enhancing data management capabilities.
  - Implemented end-to-end tests for GraphQL import/export functionalities.
  - Introduced a method for importing plugins from external JSON data, improving plugin management.
  - Enhanced scene builder functionality with export mode support for better scene management.
  - Added functionality to import storytelling data, expanding storytelling capabilities.

- **Bug Fixes**
  - Improved error handling in various import methods to ensure robustness.

- **Tests**
  - Expanded test coverage for new import/export functionalities, including tests for project and storytelling import functionality.

### CODE REVIEW

File: server/e2e/gql_import_export_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	response.Status(http.StatusOK)
+
+	// response.JSON().
+	// 	Object().
+	// 	Value("data").Object().
+	// 	Value("exportProject").Object().
+	// 	Value("projectData").Object()
+
+}
```
[the remark this answers is not in the document]
Comment (hexaforce) [reply server/e2e/gql_import_export_test.go:44]: Disabled due to problems with graphql

File: server/e2e/gql_import_export_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
```
[the remark this answers is not in the document]
Comment (hexaforce) [reply server/e2e/gql_import_export_test.go:80]: Disabled due to problems with graphql

File: server/pkg/scene/builder/decoder.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	return scene.NewWidgetArea(
+		widgetIDs,
+		parseWidgetAlign(widgetAreaJSON.Align),
+		parseWidgetAreaPadding(widgetAreaJSON.Padding),
+		widgetAreaJSON.Gap,
+		widgetAreaJSON.Centered,
+		widgetAreaJSON.Background,
+	)
+}
```
[the remark this answers is not in the document]
Comment (hexaforce) [reply server/pkg/scene/builder/decoder.go:81]: ok

File: server/pkg/scene/builder/builder.go [hunk 4]
```diff
@@ -33,15 +34,18 @@ type Builder struct {
 	nlsLayer    *nlslayer.NLSLayerList
 	layerStyles *scene.StyleList
 	story       *storytelling.Story
+
+	export bool
```
[the remark this answers is not in the document]
Comment (hexaforce) [reply server/pkg/scene/builder/builder.go:38]: Fixed in exportType

File: server/internal/adapter/gql/resolver_mutation_project.go [hunk 5]
```diff
@@ -107,3 +109,81 @@ func (r *mutationResolver) DeleteProject(ctx context.Context, input gqlmodel.Del
...
+		return nil, err
+	}
+	res["project"] = gqlmodel.ToProject(prj)
+	res["plugins"] = gqlmodel.ToPlugins(plgs)
+
+	return &gqlmodel.ImportProjectPayload{
+		ProjectData: res,
+	}, nil
+
+}
```
[the remark this answers is not in the document]
Comment (hexaforce) [reply server/internal/adapter/gql/resolver_mutation_project.go:189]: Add transaction processing

File: server/internal/adapter/gql/gqlmodel/convert.go [hunk 6]
```diff
@@ -64,6 +65,33 @@ func ToVisualizerRef(v visualizer.Visualizer) *Visualizer {
...
+	case PluginExtensionTypeStory:
+		return plugin.ExtensionTypeStory
+	case PluginExtensionTypeStoryPage:
+		return plugin.ExtensionTypeStoryPage
+	case PluginExtensionTypeStoryBlock:
+		return plugin.ExtensionTypeStoryBlock
+	default:
+		return plugin.ExtensionType("")
+	}
+}
```
Comment (pyshx) [inline server/internal/adapter/gql/gqlmodel/convert.go:93]: Please arrange them in alphabetical order.

File: server/internal/usecase/interactor/storytelling.go [hunk 7]
```diff
@@ -992,6 +997,137 @@ func (i *Storytelling) MoveBlock(ctx context.Context, inp interfaces.MoveBlockPa
 	return story, page, &inp.BlockID, inp.Index, nil
 }
 
+func (i *Storytelling) ImportStoryTelling(ctx context.Context, sceneData map[string]interface{}) (*storytelling.Story, error) {
```
Comment (pyshx) [inline server/internal/usecase/interactor/storytelling.go:1000]: Rename this to ImportStory

**Review by pyshx — APPROVED** [review]

File: server/internal/usecase/interactor/storytelling.go (hunk 7 above)
Comment (hexaforce) [reply server/internal/usecase/interactor/storytelling.go:1000]: yes

**Review by airslice — APPROVED** [review]

**Pull request merged**